caffe2/onnx/backend.cc

```diff
@@ -0,0 +1,1210 @@
+    // a newlly added one and we need to add this to argument too
+    if (!onnx_attrs_.count(kv.first)) {
+      const auto& attr = kv.second;
+      auto* arg = args.Add();
+      arg->set_name(mapper(attr.name()));
+      CopyOnnxAttrValueToCaffe2Arg(arg, attr);
+    }
+  }
+
+  return args;
+}
+
+const std::unordered_map<std::string, int>&
+Caffe2Backend::get_broken_operators() const {
+  const static std::unordered_map<std::string, int> kBrokenOperators{};
+  return kBrokenOperators;
+}
+
+// Temporary hack for RNN related operators, as we don't have C++ interface in
+// C2 to build those operators yet
+const std::unordered_set<std::string>& Caffe2Backend::get_rnn_operators() const {
+  const static std::unordered_set<std::string> kRNNOperators{
+      "LSTM", "GRU", "RNN"};
+  return kRNNOperators;
+}
+
+// Operators that are different between Caffe2 and
+// ONNX but only in their name.
+// In most cases, this should be empty - as the effort of ONNX is
+// to unify the operator definitions.
+const std::unordered_map<std::string, std::string>&
+Caffe2Backend::get_renamed_operators() const {
+  const static std::unordered_map<std::string, std::string> kRenamedOperators{
+      {"Caffe2ConvTranspose", "ConvTranspose"},
+      {"GlobalMaxPool", "MaxPool"},
+      {"GlobalAveragePool", "AveragePool"},
+      {"Pad", "PadImage"},
+      {"Neg", "Negative"},
+      {"BatchNormalization", "SpatialBN"},
+      {"InstanceNormalization", "InstanceNorm"},
+      {"MatMul", "BatchMatMul"},
+      {"Upsample", "ResizeNearest"},
+      {"Identity", "Copy"},
+      {"InstanceNormalization", "InstanceNorm"},
+      {"Equal", "EQ"},
+      {"Less", "LT"},
+      {"Greater", "GT"},
+      {"Unsqueeze", "ExpandDims"}};
+  return kRenamedOperators;
+}
+
+const std::unordered_map<std::string, std::string>&
+Caffe2Backend::get_renamed_attrs() const {
+  const static std::unordered_map<std::string, std::string> kRenamedAttrs{
+      {"kernel_shape", "kernels"}};
+  return kRenamedAttrs;
+}
+
+const std::
+    unordered_map<std::string, std::unordered_map<std::string, std::string>>&
+    Caffe2Backend::get_per_op_renamed_attrs() const {
+  const static std::
+      unordered_map<std::string, std::unordered_map<std::string, std::string>>
+          kPerOpRenamedAttrs = {{"Squeeze", {{"axes", "dims"}}},
+                                {"Unsqueeze", {{"axes", "dims"}}},
+                                {"Transpose", {{"perm", "axes"}}},
+                                {"Upsample", {{"mode", ""}}},
+                                {"ConvTranspose", {{"output_padding", "adjs"}}},
+                                {"Selu", {{"gamma", "scale"}}}};
+
+  return kPerOpRenamedAttrs;
+}
+
+// operators whose behavior is different beyond renaming
+// the value is an attribute of this class that is a
+// function from ToffeIR node_def to caffe2 op_def
+const std::unordered_map<std::string, Caffe2Backend::SpecialOpConverter>&
+Caffe2Backend::get_special_operators() const {
```

Comment:
It's better to reuse https://github.com/caffe2/caffe2/blob/master/caffe2/core/registry.h, and we can expose a register function to python side to add RNN ops converters.

Reply:
Agreed. But this is more involved. Mind if I add a follow up PR to address this?